front50-proto/src/main/proto/applications.proto

```diff
@@ -0,0 +1,241 @@
+  string owner_email = 4;
+
+  // Output only. When the application was first created.
+  google.protobuf.Timestamp create_time = 5;
+
+  // Output only. When the application was last updated.
+  google.protobuf.Timestamp update_time = 6;
+
+  // Output only. The actor which updated the application.
+  string updated_by = 7;
+
+  // A list of cloud providers that the application has configured.
+  //
+  // While not required, defining cloud providers will help inform Spinnaker's behavior for the application.
+  repeated string cloud_providers = 8;
+
+  // A bucket of named, client-typed application attributes.
+  //
+  // This allows extension of the configuration that Front50 provides for an application without modification of its
+  // core data model.
+  map<string, AttributeGroup> attribute_groups = 9;
+
+  // A named collection of instance links for an application.
+  repeated InstanceLinkGroup instance_links = 10;
+
+  // A list of all accounts an application resides in.
+  //
+  // While not required, defining accounts will help inform Spinnaker's behavior for the application.
+  repeated string accounts = 11;
+
+  // A list of application-wide notification configurations.
+  //
+  // Notifications can be triggered in response to a number of events within Spinnaker, including pipeline execution
+  // lifecycle changes.
+  repeated Notification notifications = 12;
+
+  // An arbitrary bucket of labels for the application that can be used by end-users and integrators alike.
+  map<string, string> labels = 13;
+}
+
+// AttributeGroup is a named, client-typed bucket of data used for extending Front50's data model.
+message AttributeGroup {
+  // Resource name for the attribute group (e.g. "attributeGroups/netflixTimelines")
+  string name = 1;
+
+  // The data associated with the attribute group.
+  google.protobuf.Struct data = 2;
+
+  // The required permissions for updating an attribute group.
+  repeated Permission permissions = 3;
+}
+
+// InstanceLinkGroup is a titled collection of InstanceLinks.
+message InstanceLinkGroup {
+  // The title of the group.
+  string title = 1;
+
+  // A collection of instance links.
+  repeated InstanceLink links = 2;
+}
+
+// InstanceLink is a titled URL link template for server group instances.
+message InstanceLink {
+  // The title of an instance link.
+  string title = 1;
+
+  // The link template value.
+  string link = 2;
+}
+
+// Notification is the configuration for application-wide notification recipients.
+message Notification {
+  // The type of notification (e.g. slack, email, webhook...)
+  string notification_type = 1;
+
+  // The type-specific notification recipient (e.g. email address, slack channel(s), webhook endpoint...)
+  string recipient = 2;
+
+  // A collection of event names to notify on.
+  repeated string notifying_events = 3;
+
+  // A bucket of type-specific notification config.
+  map<string, string> metadata = 6;
```

Comment:
This too could maybe be an Any, where each notification type could define their own proto message?

Using Any definitely adds some amount of overhead to the client code, but once you resolve the Any to a specific type of proto message, you get all the benefits of a proto instead of a Map. So it's kinda debatable but I'd lean towards Any personally...

Reply:
Thanks, that makes sense. When using an `Any`, is it common practice to have an associated `kind` field (in this case, `metadataKind`) as well to help resolution of the embedded proto?

Reply:
Any includes a field `type_url` to say what protobuf message it contains:
  https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/any.proto
So we could set it to things like `spinnaker.io/com.netflix.notifications.SlackMessage` or whatever. I suppose in this case, you could probably even get rid of `notification_type`, since the type of the `Any` message would tell you what kind of notification we have.

But maybe for front50 we don't want to go that route? Maybe that's shoving too much knowledge of the types into front50? Hmm...

Reply:
https://blog.envoyproxy.io/dynamic-extensibility-and-protocol-buffers-dcd0bf0b8801 has a decent description of any vs. struct. In particular, the following stick out to me:

Any:
> There is no meaningful way to make sense of the embedded proto without its schema, i.e. its protobuf descriptor.
> Messages can be efficiently (de)serialized, without JSON round-tripping.

Struct:
> This is a very flexible type and brings the advantages of dynamic typing to protobuf.
> It’s not possible to statically type check.
> This representation is not wire efficient.
> The official language specific protobuf libraries do not have first-class support for converting between Struct and arbitrary protobuf message types, instead it’s always necessary to round-trip via JSON (de)serialization ops.

Seems like the tradeoff is around how opaque the types should be from Front50's perspective. Should front50 be storing opaque blobs? Or should we require that it is extended to know about additional Message types. Based on the envoy blog post above, seems like it might be case-by-case, depending on if the API is considered an opaque extension point (biased towards a Struct) or a standard component that Front50 may need to interpret in the future (arguing for an Any).

Reply:
Yeah, good summary, thanks. For a higher-level service like Echo that deals more directly with notifications, I definitely would not expect it to be dealing in structs, and Any probably makes a lot more sense. But for a data service like front50, those considerations change lot, and maybe Any imposes too much structure...

Reply:
OK, that was helpful! Sounds like Struct is the better fit basically across the board for front50 at this point.

Reply:
Given that `front50` is going to be the owner of the data (rather than just a pass-through) I'd probably vote for stronger typing and going with the `Any`.  If it raises the bar to adding a new notification type, I actually think that might be a good thing (as that bar would be raised by forcing more structure and better documentation).

My opinion is probably strongly affected by spending a lot of time wondering "what is in this map" for intra-service functions, but I'd love to try as much as possible to avoid that for things in the API.  That being said, this lack of typing is limited to one small corner of the application object, so maybe it's reasonable here...

Reply:
I don't feel like I have enough experience with systems like this to judge where we'll hit future hurdles. How responsible is Front50 for ensuring its own data integrity? At Airbnb, we are currently extending Gate, Echo, Kayenta, Orca, and Clouddriver. We are not currently extending Front50. This is nice from an overhead perspective, but it also means that Front50 doesn't really have the knowledge to:

1. enforce any correctness when another service stores a notification type (if two systems have different ideas about the schema they want, we lose the value of typing here).
2. migrate data from one schema to another when performing major version (non-backwards compatible) upgrades.

Maybe we should be aiming to make it easy to include new Message types as plugins for each service instead of optimizing for maximal flexibility at runtime? I feel like I'd rather that flow be super easy than have to deal with all the pain around adding manual schema-checking logic on the echo side for each notification type.

Reply:
I don't disagree, but for conversation sake: I'd like to reduce use cases where we need to modify multiple services to add a new thing (cloud providers, notifications, triggers, etc). If we turn this into an Any, does front50 actually need to know the internal type? I'd love a world where Echo defines a notification and all the structure that it has, but not need to have front50 change to store that structure.

Reply:
If we're saying this is Echo-owned data that's opaque to front50, I still think storing it as an Any makes more sense than a struct. Just serialize it into a byte[] and store that. Eventually Echo is going to be using protos for its internal models too, so using protos instead of maps now just inches it towards that future.

Reply:
Ah, I think you are right @plumpy. In this case, we want the message to be opaque _except_ for the writers/readers. Because this can be completely opaque to Front50, using Any makes a lot of sense.

To clarify what I'm thinking, in Echo, we would have to have the polymorphic serialization/deserialization code, based on https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/any.proto#L60-L65:

```
// pack the slack notification spec.
SlackNotification notification = ...;
myApplication.data.put("aSlackNotifier", Any.pack(foo));

// use the deserialized form.
Any notification = myApplication.data.get("notifier")
if (any.is(SlackNotification.class)) {
  SlackNotification notification = any.unpack(Notification.class);
  notification.send(context)
}
```

Within Front50, we would never try to unpack it. Because we don't try to deserialize the encompassed envelope, we wouldn't hit any class definition errors. So in this case, Echo defines the type, uses the type, and gets all the safety involved, Front50 handles it without ever touching it.

Migrating schemas forwards still seems difficult, but easier since we at least can automatically detect backwards-incompatible schema changes.

Based on my current understanding, strongly on the `Any` side.

Reply:
For schema migration, I think you'd only do it in extreme cases. I think you could generally handle it the same as if Echo is talking to a client compiled against an old version of its protos. i.e. you added a field in Echo 2.3, but the client is using the protos from Echo 2.2. Usually you want to support that case too, although I suppose there are times when it might not be possible. Loading serialized data from an old Echo is basically the same scenario.